source/grid/grid_tools.cc

```diff
@@ -4451,6 +4452,299 @@ namespace GridTools
+      {
+        has_cells_with_more_than_dim_faces_on_boundary = false;
+
+        for (auto cell: tria.active_cell_iterators())
+          {
+            unsigned int boundary_face_counter = 0;
+            for (unsigned int f=0; f<GeometryInfo<dim>::faces_per_cell; ++f)
+              if (cell->face(f)->at_boundary())
+                boundary_face_counter++;
+            if (boundary_face_counter > dim)
+              {
+                has_cells_with_more_than_dim_faces_on_boundary = true;
+                break;
+              }
+            else if (boundary_face_counter == dim)
+              has_cells_with_dim_faces_on_boundary = true;
+          }
+        if (has_cells_with_more_than_dim_faces_on_boundary)
+          {
+            tria.refine_global(1);
+            refinement_cycles++;
+          }
+      }
+
+    if (has_cells_with_dim_faces_on_boundary)
+      {
+        tria.refine_global(1);
+        refinement_cycles++;
+      }
+    else
+      {
+        return;
+      }
+
+    std::vector<bool> cells_to_remove(tria.n_active_cells(), false);
+    std::vector<Point<spacedim> > vertices = tria.get_vertices();
+
+    std::vector<bool> faces_to_remove(tria.n_raw_faces(),false);
+
+    std::vector<CellData<dim> > cells_to_add;
+    SubCellData                 subcelldata_to_add;
+
+    // Trick compiler for dimension independent things
+    const unsigned int
+    v0 = 0, v1 = 1,
+    v2 = (dim > 1 ? 2:0), v3 = (dim > 1 ? 3:0),
+    v4 = (dim > 2 ? 4:0), v5 = (dim > 2 ? 5:0),
+    v6 = (dim > 2 ? 6:0), v7 = (dim > 2 ? 7:0);
+
+    for (auto cell : tria.active_cell_iterators())
+      {
+        double angle_fraction = 0;
+        unsigned int vertex_at_corner = numbers::invalid_unsigned_int;
+
+        if (dim == 2)
+          {
+            Tensor<1,spacedim> p0;
+            p0[spacedim > 1 ? 1 : 0] = 1;
+            Tensor<1,spacedim> p1;
+            p1[0] = 1;
+
+            if (cell->face(v0)->at_boundary() && cell->face(v3)->at_boundary())
+              {
+                p0 = cell->vertex(v0) -  cell->vertex(v2);
+                p1 = cell->vertex(v3) -  cell->vertex(v2);
+                vertex_at_corner = v2;
+              }
+            else  if (cell->face(v3)->at_boundary() && cell->face(v1)->at_boundary())
+              {
+                p0 = cell->vertex(v2) -  cell->vertex(v3);
+                p1 = cell->vertex(v1) -  cell->vertex(v3);
+                vertex_at_corner = v3;
+              }
+            else  if (cell->face(1)->at_boundary() && cell->face(2)->at_boundary())
+              {
+                p0 = cell->vertex(v0) -  cell->vertex(v1);
+                p1 = cell->vertex(v3) -  cell->vertex(v1);
+                vertex_at_corner = v1;
+              }
+            else  if (cell->face(2)->at_boundary() && cell->face(0)->at_boundary())
+              {
+                p0 = cell->vertex(v2) -  cell->vertex(v0);
+                p1 = cell->vertex(v1) -  cell->vertex(v0);
```

Comment:
I hope that we can replace these cases with more generic code. Let me think about it (probably tomorrow).